### PR TITLE
New logging 

### PR DESCRIPTION
**Pull request by PhilipVinc**
This is my proposal for the new logging system, going alongside nk3

Things I disliked:
 - The output format previously was a json file with a single dict with a key, `output`, with a list of data logged at every iteration, and another key `iterations` with the iteration numbers.
 - To deserialise this format one has to determine the objects logged inside every iterations (some might not always be logged, such as objservables), initialise some accumulators, iterate through every element in `output` and add the right object to the right accumulator. Also, one has then to reconstruct the iteration numbers for objects that are not logged at every iteration

```
{`Output`: [
    { 'Energy': {'Mean': 0.3, 'Variance':....} ,
    { 'Energy': {'Mean': 0.2, 'Variance':....} ,
    { 'Energy': {'Mean': 0.1, 'Variance':....} ,
...
   ],
`Iters`:[1,2,3...]
}
```
Changes:
 - The new output format is still a json file, but now it's simply a dict of time series : 

```
{ 'Energy': {
    'Mean': {'iters':[1,2,3...] , 'values':[0.3, 0.2,0.1....]
    'Variance': {'iters':[1,2,3...] , 'values':[0.01, 0.02,0.01....]
    'Sigma': {'iters':[1,2,3...] , 'values':[0.01, 0.02,0.01....]
}
}
```

this makes it much easier to de-serialize, as simply deserialising with json gives a structure that can be plotted with:
```python
import Matplotlib.pyplot as plt

data = json.load(open("test.log"))

plt.plot(data['Energy']['Mean']['iters'], data['Energy']['Mean']['values']
```

internally, during runtime, all logged quantities are stored in a tree of `History` objects (time-series), that store both the iteration and the values at every iteration. They are then natively serialised to json...

### CODE REVIEW

File: netket/logging/json_log.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            starting_json_content = json.load(open(output_prefix + ".log"))
+
+        elif file_exists and mode == "fail":
+            raise ValueError(
+                "Output file already exists. Either delete it manually or change `output_prefix`."
+            )
+
+        dir_name = _path.dirname(output_prefix)
+        if dir_name is not "":
+            os.makedirs(dir_name, exist_ok=True)
```
Comment (femtobit) [inline netket/logging/json_log.py:103]: (Got a warning for this when running the example.)
```suggestion
        if dir_name != "":
            os.makedirs(dir_name, exist_ok=True)
```

**Review by femtobit — COMMENTED** [review]
Looks like a good idea to me.

Is it necessary to have `iters` repeated for all fields? I can see some justification in adding it for the observables - they could be estimated at different intervals (e.g., when some are more costly to evaluate). But shouldn't `Mean`, `Variance`, etc. always be computed at the same iterations?

It doesn't really affect the ease of plotting: `plot(data["Energy"]["iters"], data["Energy"]["Mean"])`.

**Comment by PhilipVinc** [comment]
It's a consequence  of the fact that every logged quantity is a TimeSeries (aka `History`), and every one of those stores the iterations...

The problem is that `Stats` are de-facto converted to a dict of their 5 fields, Mean, Variance, R_hat, etc and each one of those is a time series.

If you have better ideas, I'm all ears

**Comment by PhilipVinc** [comment]
Ideally, I'd like to provide a small utility to de-serialize those json to `History` objects. 
In that way, it becomes easier to plot stuff because you can just do

```python
plot(data["Energy"]["Mean"].iters, data["Energy"]["Mean"], data["Energy"]["variance"]).
```
and unless specified it always takes the values...

**Comment by PhilipVinc** [comment]
my idea is, in theory, yes.

**Comment by PhilipVinc** [comment]
It should work as discussed now.
Basically I implemented a MultiValue History object, that takes in several keys from a compound object( aka, Stats, or anything that has a to_compound() method)

**Pull request merged**